src/realizations/catchment/Bmi_Py_Formulation.cpp

```diff
@@ -7,10 +7,154 @@ using namespace models::bmi;
+    return get_bmi_model()->GetInputVarNames();
+}
+
+const vector<string> Bmi_Py_Formulation::get_bmi_output_variables() {
+    return get_bmi_model()->GetOutputVarNames();
+}
+
 std::string Bmi_Py_Formulation::get_formulation_type() {
     return "bmi_py";
 }
 
+string Bmi_Py_Formulation::get_output_line_for_timestep(int timestep, std::string delimiter) {
+    // TODO: something must be added to store values if more than the current time step is wanted
+    // TODO: if such a thing is added, it should probably be configurable to turn it off
+    if (timestep != (next_time_step_index - 1)) {
+        throw std::invalid_argument("Only current time step valid when getting output for BMI Python formulation");
+    }
+    std::string output_str;
+
+    for (const std::string& name : get_output_variable_names()) {
+        output_str += (output_str.empty() ? "" : ",") + std::to_string(get_var_value_as_double(name));
+    }
+    return output_str;
+}
+
+double Bmi_Py_Formulation::get_response(time_step_t t_index, time_step_t t_delta) {
+    if (get_bmi_model() == nullptr) {
+        throw std::runtime_error("Trying to process response of improperly created BMI C formulation.");
+    }
+    if (t_index < 0) {
+        throw std::invalid_argument("Getting response of negative time step in BMI C formulation is not allowed.");
+    }
+    // Use (next_time_step_index - 1) so that second call with current time step index still works
+    if (t_index < (next_time_step_index - 1)) {
+        // TODO: consider whether we should (optionally) store and return historic values
+        throw std::invalid_argument("Getting response of previous time step in BMI C formulation is not allowed.");
+    }
+
+    // The time step delta size, expressed in the units internally used by the model
+    double t_delta_model_units = get_bmi_model()->convert_seconds_to_model_time((double)t_delta);
+    if (next_time_step_index <= t_index) {
+        double model_time = get_bmi_model()->GetCurrentTime();
+        // Also, before running, make sure this doesn't cause a problem with model end_time
+        if (!get_allow_model_exceed_end_time()) {
+            int total_time_steps_to_process = abs((int)t_index - next_time_step_index) + 1;
+            if (get_bmi_model()->GetEndTime() < (model_time + (t_delta_model_units * total_time_steps_to_process))) {
+                throw std::invalid_argument("Cannot process BMI C formulation to get response of future time step "
+                                            "that exceeds model end time.");
+            }
+        }
+    }
+
+    while (next_time_step_index <= t_index) {
+        double model_initial_time = get_bmi_model()->GetCurrentTime();
+        set_model_inputs_prior_to_update(model_initial_time, t_delta);
+        if (t_delta_model_units == get_bmi_model()->GetTimeStep())
+            get_bmi_model()->Update();
+        else
+            get_bmi_model()->UpdateUntil(model_initial_time + t_delta_model_units);
```

Comment:
Just to note, UpdateUntil may not always be implemented to support this type of updating until a specific time...